drivers/comparator/comparator_stm32_comp.c

```diff
@@ -0,0 +1,324 @@
+		return 1;
+	}
+	return 0;
+}
+
+static int stm32_comp_set_trigger_callback(const struct device *dev, comparator_callback_t callback,
+					   void *user_data)
+{
+	const struct stm32_comp_config *cfg = dev->config;
+	struct stm32_comp_data *data = dev->data;
+
+	irq_disable(cfg->irq_nr);
+
+	data->callback = callback;
+	data->user_data = user_data;
+
+	irq_enable(cfg->irq_nr);
+
+	if (data->callback != NULL && stm32_comp_trigger_is_pending(dev)) {
+		callback(dev, user_data);
+	}
+
+	return 0;
+}
+
+static DEVICE_API(comparator, stm32_comp_comp_api) = {
+	.get_output = stm32_comp_get_output,
+	.set_trigger = stm32_comp_set_trigger,
+	.set_trigger_callback = stm32_comp_set_trigger_callback,
+	.trigger_is_pending = stm32_comp_trigger_is_pending,
+};
+
+static int stm32_comp_pm_callback(const struct device *dev, enum pm_device_action action)
+{
+	const struct stm32_comp_config *cfg = dev->config;
+	COMP_TypeDef *comp = cfg->comp;
+
+	if (action == PM_DEVICE_ACTION_RESUME) {
+		if (cfg->lock_enable) {
+			LL_COMP_Lock(comp);
+		}
+		LL_COMP_Enable(comp);
+	}
+
+#ifdef CONFIG_PM_DEVICE
+	if (action == PM_DEVICE_ACTION_SUSPEND) {
+		LL_COMP_Disable(comp);
+	}
+#endif
+
+	return 0;
+}
+
+static void stm32_comp_irq_handler(const struct device *dev)
+{
+	const struct stm32_comp_config *cfg = dev->config;
+	struct stm32_comp_data *data = dev->data;
+
+	if (stm32_exti_is_pending(cfg->exti_line_number)) {
+		stm32_exti_clear_pending(cfg->exti_line_number);
+	}
+
+	if (data->callback == NULL) {
+		return;
+	}
+
+	data->callback(dev, data->user_data);
+}
+
+static int stm32_comp_init(const struct device *dev)
+{
+	const struct device *const clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	const struct stm32_comp_config *cfg = dev->config;
+	COMP_TypeDef *comp = cfg->comp;
+	int ret = 0;
+
+	if (!device_is_ready(clk)) {
+		LOG_ERR("%s clock control device not ready", dev->name);
+		return -ENODEV;
+	}
+
+	/* Enable COMP bus clock */
+	ret = clock_control_on(clk, &cfg->pclken[0]);
+	if (ret != 0) {
+		LOG_ERR("%s clock op failed (%d)", dev->name, ret);
+		return ret;
+	}
+
+	/* Enable COMP clock source */
+	clock_control_configure(clk, &cfg->pclken[1], NULL);
+	if (ret != 0) {
+		LOG_ERR("%s clock configure failed (%d)", dev->name, ret);
+		return ret;
+	}
+
+	/* Configure COMP inputs as specified in Device Tree, if any */
+	ret = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0 && ret != -ENOENT) {
+		/*
+		 * If the COMP is used only with internal channels, then no pinctrl is
+		 * provided in Device Tree, and pinctrl_apply_state returns -ENOENT,
+		 * but this should not be treated as an error.
+		 */
+		LOG_ERR("%s pinctrl setup failed (%d)", dev->name, ret);
+		return ret;
+	}
+
+	if (LL_COMP_IsLocked(comp)) {
+		/* COMP instance shall not be locked */
+		LOG_ERR("%s COMP instance is locked", dev->name);
+		return -EACCES;
+	}
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_comp)
+	LL_COMP_SetPowerMode(comp, cfg->power_mode);
```

Comment:
Would you be OK to use `#ifndef CONFIG_SOC_SERIES_STM32G4X` instead, it would better scale to other STM32 SoCs that all but G4xx support `LL_COMP_SetPowerMode()`.

Reply:
@bjarki-andreasen @etienne-lms I've added @etienne-lms's suggestion so far. Please agree on an alternative - I will change again then.